File: backend/aci_integration/routes/mcp_tools.py
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, ValidationError
from typing import Any, Dict, List, Optional

# Assuming the ACI server code is structured to be imported
# Need to confirm the exact import path and function names
# from ....aci.server import search_tools, execute_tool, configure_tool, register_tool

router = APIRouter()

class ToolSearchResponse(BaseModel):
    """Response model for tool search."""
    tools: List[Dict[str, Any]] # Placeholder, need to define a proper Tool model

class ToolExecuteRequest(BaseModel):
    """Request model for tool execution."""
    parameters: Dict[str, Any]

class ToolExecuteResponse(BaseModel):
    """Response model for tool execution."""
    result: Any # Placeholder, need to define a proper execution result model

class ToolConfigureRequest(BaseModel):
    """Request model for tool configuration."""
    tool_type: str
    configuration_data: Dict[str, Any]

class ToolRegisterRequest(BaseModel):
    """Request model for tool registration."""
    tool_type: str
    configuration_identifier: Optional[str] = None
    configuration_data: Optional[Dict[str, Any]] = None

@router.get("/api/v1/tools/search", response_model=ToolSearchResponse)
async def search_mcp_tools(
    query: Optional[str] = Query(None, description="Search query for tools")
):
    """
    Searches for available ACI MCP tools.

    Includes basic validation for the 'query' parameter using FastAPI's Query.
    Implements error handling for potential issues during the search operation
    by the embedded ACI code.
    """
    try:
        # TODO: Call the embedded ACI MCP code to search for tools
        # Example: tools = await search_tools(query)
        # For now, return a placeholder
        print(f"Searching for tools with query: {query}")
        placeholder_tools = [
            {"name": "example_tool_1", "description": "This is an example tool."},
            {"name": "example_tool_2", "description": "Another example tool."}
        ]
        # Simulate a potential error from the ACI layer, e.g., if the query is too complex
        if query and "complex_error" in query:
             raise ValueError("Simulated complex query error from ACI")

        return ToolSearchResponse(tools=placeholder_tools)
    except ValueError as e:
        # Handle specific known errors originating from the ACI layer or input validation
        raise HTTPException(status_code=400, detail=f"Invalid search query or ACI error: {e}")
    except Exception as e:
        # Catch any other unexpected errors during the search operation
        raise HTTPException(status_code=500, detail=f"Internal server error during tool search: {e}")

@router.post("/api/v1/tools/{tool_name}/execute", response_model=ToolExecuteResponse)
async def execute_mcp_tool(
    tool_name: str,
    request: ToolExecuteRequest
):
    """
    Executes a specific ACI MCP tool.

    Uses Pydantic model `ToolExecuteRequest` for request body validation.
    Includes error handling for potential issues during tool execution
    by the embedded ACI code, such as tool not found or execution errors.
    """
    # FastAPI automatically validates the request body against ToolExecuteRequest
    try:
        # TODO: Call the embedded ACI MCP code to execute the tool
        # Example: execution_result = await execute_tool(tool_name, request.parameters)
        # For now, return a placeholder
        print(f"Executing tool '{tool_name}' with parameters: {request.parameters}")
        # Simulate potential errors from the ACI layer
        if tool_name == "non_existent_tool":
            raise FileNotFoundError(f"Tool '{tool_name}' not found in ACI")
        if "simulated_execution_error" in request.parameters:
             raise RuntimeError("Simulated execution error from ACI")

        placeholder_result = {"status": "success", "output": f"Executed {tool_name}"}
        return ToolExecuteResponse(result=placeholder_result)
    except FileNotFoundError as e:
        # Handle case where the tool is not found in the ACI layer
        raise HTTPException(status_code=404, detail=str(e))
    except (ValueError, TypeError, RuntimeError) as e:
        # Handle specific errors related to execution parameters or ACI runtime issues
        raise HTTPException(status_code=400, detail=f"Tool execution failed: {e}")
    except Exception as e:
        # Catch any other unexpected errors during execution
        raise HTTPException(status_code=500, detail=f"Internal server error during tool execution: {e}")

@router.post("/api/v1/tools/configure")
async def configure_mcp_tool(
    request: ToolConfigureRequest
):
    """
    Configures an ACI MCP tool.

    Uses Pydantic model `ToolConfigureRequest` for request body validation.
    Includes error handling for potential issues during tool configuration
    by the embedded ACI code, such as invalid configuration data.
    """
    # FastAPI automatically validates the request body against ToolConfigureRequest
    try:
        # TODO: Call the embedded ACI MCP code to submit the configuration
        # Example: configuration_result = await configure_tool(request.tool_type, request.configuration_data)
        # For now, return a placeholder
        print(f"Configuring tool '{request.tool_type}' with data: {request.configuration_data}")
        # Simulate a potential error from the ACI layer, e.g., invalid configuration data format
        if not request.configuration_data or "invalid_format" in request.configuration_data:
             raise ValueError("Simulated invalid configuration data format from ACI")

        # Assuming a simple success response for now
        return {"status": "success", "message": f"Tool '{request.tool_type}' configured."}
    except ValueError as e:
        # Handle specific known errors related to configuration data or ACI configuration issues
        raise HTTPException(status_code=400, detail=f"Tool configuration failed: {e}")
    except Exception as e:
        # Catch any other unexpected errors during configuration
        raise HTTPException(status_code=500, detail=f"Internal server error during tool configuration: {e}")

@router.post("/api/v1/tools/register")
async def register_mcp_tool(
    request: ToolRegisterRequest
):
    """
    Registers an ACI MCP tool.

    Uses Pydantic model `ToolRegisterRequest` for request body validation.
    Includes error handling for potential issues during tool registration
    by the embedded ACI code, such as tool type already exists.
    """
    # FastAPI automatically validates the request body against ToolRegisterRequest
    try:
        # TODO: Call the embedded ACI MCP code to initiate registration
        # Example: registration_status = await register_tool(request.tool_type, request.configuration_identifier, request.configuration_data)
        # For now, return a placeholder
        print(f"Registering tool '{request.tool_type}' with identifier: {request.configuration_identifier} and data: {request.configuration_data}")
        # Simulate a potential error from the ACI layer, e.g., tool_type already exists
        if request.tool_type == "existing_tool_type":
             raise ValueError(f"Tool type '{request.tool_type}' already exists in ACI")

        # Assuming a simple success response for now
        return {"status": "registration_initiated", "tool_type": request.tool_type}
    except ValueError as e:
        # Handle specific known errors related to registration data or ACI registration issues
        raise HTTPException(status_code=400, detail=f"Tool registration failed: {e}")
    except Exception as e:
        # Catch any other unexpected errors during registration
        raise HTTPException(status_code=500, detail=f"Internal server error during tool registration: {e}")

File: frontend/src/api/aci/mcpTools.ts
// frontend/src/api/aci/mcpTools.ts

// Placeholder types - replace with actual types based on backend design
export interface McpTool {
  id: string;
  name: string;
  description: string;
  // Add other relevant tool fields
}

export interface McpToolExecutionResult {
  // Define structure based on backend execution result
  result: any; // Placeholder
}

export interface McpToolConfiguration {
  // Define structure based on backend configuration
  configuration: any; // Placeholder
}

/**
 * Searches for available MCP Tools.
 * @param query Optional search query.
 * @returns A promise resolving to an array of McpTool.
 */
export async function searchMcpTools(query?: string): Promise<McpTool[]> {
  try {
    const url = query ? `/api/aci/mcp/tools?q=${encodeURIComponent(query)}` : '/api/aci/mcp/tools';
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data: McpTool[] = await response.json();
    return data;
  } catch (error) {
    console.error("Error searching MCP tools:", error);
    throw error;
  }
}

/**
 * Executes a specific MCP Tool.
 * @param toolId The ID of the tool to execute.
 * @param args The arguments for the tool execution.
 * @returns A promise resolving to the execution result.
 */
export async function executeMcpTool(toolId: string, args: any): Promise<McpToolExecutionResult> {
  try {
    const response = await fetch(`/api/aci/mcp/tools/${toolId}/execute`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(args),
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data: McpToolExecutionResult = await response.json();
    return data;
  } catch (error) {
    console.error(`Error executing MCP tool ${toolId}:`, error);
    throw error;
  }
}

/**
 * Configures a specific MCP Tool.
 * @param toolId The ID of the tool to configure.
 * @param configuration The configuration data.
 * @returns A promise resolving when the configuration is successful.
 */
export async function configureMcpTool(toolId: string, configuration: McpToolConfiguration): Promise<void> {
  try {
    const response = await fetch(`/api/aci/mcp/tools/${toolId}/configure`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(configuration),
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
  } catch (error) {
    console.error(`Error configuring MCP tool ${toolId}:`, error);
    throw error;
  }
}

/**
 * Registers a new MCP Tool.
 * @param toolData The data for the tool to register.
 * @returns A promise resolving to the registered McpTool.
 */
export async function registerMcpTool(toolData: any): Promise<McpTool> {
  try {
    const response = await fetch('/api/aci/mcp/tools', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(toolData),
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data: McpTool = await response.json();
    return data;
  } catch (error) {
    console.error("Error registering MCP tool:", error);
    throw error;
  }
}